core/jobs/batch_jobs/state_interaction_validation_jobs.py

```diff
@@ -0,0 +1,1382 @@
+                text_input_interaction_values = []
+
+                # Text input height >= 1 and <= 10.
+                rows_value = int(
+                    state.interaction.customization_args['rows'].value
+                )
+                if rows_value < 1 or rows_value > 10:
+                    text_input_interaction_values.append(
+                        f'Rows having value {rows_value} is either '
+                        f'less than 1 or greater than 10.'
+                    )
+
+                # Contains should come after other rule where its a substring.
+                text_input_interaction_values += (
+                    ExpStateInteractionValidationJob.
+                    _contains_should_come_after(state)
+                )
+
+                # Starts with should come after other rule where its a prefix.
+                text_input_interaction_values += (
+                    ExpStateInteractionValidationJob.
+                    _starts_with_should_come_after(state)
+                )
+
+                if len(text_input_interaction_values) > 0:
+                    states_with_values.append(
+                        {
+                            'state_name': state_name,
+                            'text_input_interaction_values': (
+                                text_input_interaction_values)
+                        }
+                    )
+
+        return states_with_values
+
+    invalid_end_interac = []
+
+    @staticmethod
+    def filter_invalid_end_interac(exp, exp_id_list):
+        """Filter invalid end interaction states.
+
+        Args:
+            exp: exp_domain.Exploration. The exploration.
+            exp_id_list: List[str]. List of all exploration ids.
+
+        Returns:
+            found_invalid: bool. Returns True if invalid exploration.
+        """
+        states = exp.states
+        found_invalid = False
+        exp_end_interac_values = []
+        for state_name, state in states.items():
+            if state.interaction.id == 'EndExploration':
+                invalid_state_exp_ids = []
+                recc_exp_ids = (
+                    state.interaction.customization_args
+                    ['recommendedExplorationIds'].value
+                )
+
+                for exp_id in recc_exp_ids:
+                    found_invalid = True
+                    if exp_id not in exp_id_list:
+                        invalid_state_exp_ids.append(exp_id)
+                if len(invalid_state_exp_ids) > 0:
+                    (
+                        exp_end_interac_values.append(
+                            {'state_name': state_name,
+                            'invalid_exps': invalid_state_exp_ids}
+                        )
+                    )
+        if len(exp_end_interac_values) > 0:
+            ExpStateInteractionValidationJob.invalid_end_interac.append(
+                {exp.id: exp_end_interac_values}
+            )
```

Comment:
Actually, I and Eric were having a discussion regarding this part. The main reason why we had to opt for this method was, that whenever we tried to pass the collections as a list to any other functions or map functions we wouldn't be able to access the list we got an error and when we tried passing in the `filter` function it was working fine but now the problem was to store the error so that's why we created a global variable to store.

Reply:
Sorry, I will probably need a more precise explanation of what exactly didn't work.

Reply:
Sure,
Previously I was doing something like this to pass the PCollection as a list
```
self.some_filter_function(exp_states, exp_list=beam.pvalue.AsIter(exp_ids_pcoll))
```
also tried, please note that the variable `exp_ids_pcoll` is the PCollection of exp ids
```
self.some_filter_function(exp_states, exp_list=beam.pvalue.AsList(exp_ids_pcoll))
```
I received this error `'_InvalidUnpickledPCollection' object has no attribute 'windowing'` which I tried to

After that I tried using `beam.combiners.ToList()` to the PCollection and then passing to the function then I received this error `_InvalidUnpickledPCollection' object is not subscriptable` while accessing the list.

Thanks
@vojtechjelinek

Reply:
Done!
PTAL